gax/src/main/java/com/google/api/gax/batching/v2/BatcherImpl.java

```diff
@@ -74,27 +87,32 @@ public BatcherImpl(
       BatchingDescriptor<ElementT, ElementResultT, RequestT, ResponseT> batchingDescriptor,
       UnaryCallable<RequestT, ResponseT> unaryCallable,
       RequestT prototype,
-      BatchingSettings batchingSettings) {
+      BatchingSettings batchingSettings,
+      ScheduledExecutorService executor) {
     this.batchingDescriptor =
         Preconditions.checkNotNull(batchingDescriptor, "batching descriptor cannot be null");
     this.unaryCallable = Preconditions.checkNotNull(unaryCallable, "callable cannot be null");
     this.prototype = Preconditions.checkNotNull(prototype, "request prototype cannot be null");
     this.batchingSettings =
         Preconditions.checkNotNull(batchingSettings, "batching setting cannot be null");
+    this.executor = Preconditions.checkNotNull(executor, "executor cannot be null");
+    this.maxDelay = batchingSettings.getDelayThreshold();
   }
 
   /** {@inheritDoc} */
   @Override
   public ApiFuture<ElementResultT> add(ElementT element) {
     Preconditions.checkState(!isClosed, "Cannot add elements on a closed batcher");
+    SettableApiFuture<ElementResultT> result = SettableApiFuture.create();
 
     if (currentOpenBatch == null) {
       currentOpenBatch = new Batch<>(prototype, batchingDescriptor, batchingSettings);
+      currentOpenBatch.add(element, result);
+      executor.schedule(pushCurrentBatchRunnable, maxDelay.toMillis(), TimeUnit.MILLISECONDS);
+    } else {
+      currentOpenBatch.add(element, result);
```

Comment:
How else can I refactor it to reduce this duplicate code... all suggestions are welcome.
(some background info: `executor.schedule(pushCurrentBatchRunnable, maxDelay...` can auto trigger it and can make `currentOpenBatch` `null`).